### DISCUSSION TITLE
How to validation for complex property?

### DISCUSSION THREAD

**Discussion by GF-Huang** [canton7/Stylet#236] [Q&A]
For example:
```cs
public abstract class Animal : PropertyChangedBase { 
    public int Weight { get; set; }
}

public Cat : Animal { 
    public CatBreedsEnum CatBreeds { get; set; }
}

public Dog: Animal { 
    public DogBreedsEnum DogBreeds { get; set; }
}

public AnimalValidator : FluentValidation.AbstractValidator<Animal> {
    public AnimalValidator() => RuleFor(a => a.Weight).InclusiveBetween(1, 50);
}

public CatValidator : FluentValidation.AbstractValidator<Cat> {
    public CatValidator(AnimalValidator validator) { 
        Include(validator);
        RuleFor(c => c.CatBreeds).Equal(CatBreedsEnum.DragonLi); 
    }
}

public DogValidator : FluentValidation.AbstractValidator<Dog> {
    public DogValidator(AnimalValidator validator) { 
        Include(validator);
        RuleFor(d => d.DogBreeds).Equal(DogBreedsEnum.XXX); 
    }
}

public class ViewModel : ValidatingModelBase {
    public Animal MyPet { get; set; }

    public ViewModel(IModelValidator<ViewModel> validator) : base(validator) { }

    public void Save() {
        if (Validator()) // always true even if MyPet.Weight large than 50
            // ...
    }
}

public ViewModelValidator : FluentValidation.AbstractValidator<ViewModel> {
    public ViewModelValidator(CatValidator catValidator, DogValidator dogValidator) {
        RuleFor(vm => vm.MyPet).SetInheritanceValidator(v => {
             v.Add(catValidator);
             v.Add(dogValidator);
        });
    }
}

```

```xaml
...
<TextBox Text="{Binding MyPet.Weight}" />
...
```

The other scaffold codes I make like https://github.com/canton7/Stylet/wiki/ValidatingModelBase. I'm not going to write it out.

The auto validation not work and `Validate()` always return `True`.

Do I need to modify `Animal : PropertyChangedBase` to `Animal : ValidatingModelBase` and set `Validator` by constructor?

**Comment by canton7** [comment]
What does `GetErrors("MyPet")` return?

**Reply by GF-Huang** [reply]
Sorry, I don't know. This is just a minimalist example to illustrate my problem, the actual code is quite complex. Could you give me some complex property validation example which use `FluentValidation`?

**Reply by GF-Huang** [reply]
And I had edit the example by adding `xaml`, maybe it has something to do with this binding, right?

**Reply by canton7** [reply]
Since you put the example together, presumably you put together a small project to check that it still replicates your problem? Just run the method I mentioned there. Alternatively, just replace "MyPet" with the name of one of your properties which has errors (but you don't think it should have errors) to see what those errors are.

I'm not an expert on FluentValidation -- Stylet calls `ValidatePropertyAsync` and `ValidateAllPropertiesAsync`. I suspect that FluentValidation is returning an error to you, and calling `GetErrors` will see what that error is. Once we've confirmed that Stylet's forwarding errors from FluentValidation correctly, I'm afraid it's between you and FluentValidation.

**Reply by GF-Huang** [reply]
Will Stylet calls `ValidatePropertyAsync` and `ValidateAllPropertiesAsync` on `MyPet` even if `MyPet` is not a `ValidatingModelBase`?

**Reply by canton7** [reply]
It calls those methods on the IModelValidator

**Reply by GF-Huang** [reply]
[ValidationDemo.zip](https://github.com/canton7/Stylet/files/6194900/ValidationDemo.zip)

Here is demo project, could you help me to take a look why there is no red box around the `TextBox` even if validate failed?

**Reply by canton7** [reply]
![image](https://user-images.githubusercontent.com/568104/112279638-0acd6b80-8c7c-11eb-8027-dcde4b2a01e8.png)

Your validator returned a validation error for the property "MyPet.Weight". WPF doesn't understand that: it expects the property to be the name of a property on the object being validated. Since ShellViewModel is the object being validated, WPF expects the property names to be e.g. "MyPet".

**Reply by GF-Huang** [reply]
But is there any ways to solves this? In other words, make nested property validation works well and show red border around `TextBox` if there's error.

**Reply by canton7** [reply]
This is between you and WPF I'm afraid -- Stylet implements [`INotifyDataErrorInfo`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo?view=net-5.0), and reports back the same property errors which your validator gives it. Maybe you can add some logic to `FluentModelValidator` so that all errors are truncated to the first `.`, so that `MyPet.Weight` gets turned into `MyPet`?
